File: leetcode/python/102-levelOrder.py
class TreeNode:
    def __init__(self, val, left = None, right = None):
        self.val = val
        self.left = left
        self.right = right
    

from typing import Optional, List

class Solution:
    def levelOrder(self, root: Optional[TreeNode]) -> List[List[int]]:
        res = [];
        if not root:
            return res
        def bfs (node, level):
            if not node:
                return
    
            if len(res) <= level - 1:
                res.append([])

            res[level - 1].append(node.val)

            bfs(node.left, level + 1)
            bfs(node.right, level + 1)

        bfs(root, 1)
        return res

node0 = TreeNode(1)
node1 = TreeNode(2)
node2 = TreeNode(3, node0, node1)
node3 = TreeNode(4, node2)
node4 = TreeNode(5, node3)

print(Solution().levelOrder(node4))

File: leetcode/javascript/102-levelOrder.js
/**
 * 给你二叉树的根节点 root ，返回其节点值的 层序遍历 。 （即逐层地，从左到右访问所有节点）。
 * 
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */

/**
 * @description 二叉树的层序遍历
 * @param {TreeNode} root
 * @return {number[][]}
 */
var levelOrder = function(root) {
    const res = [];

    if (!root) {
        return [];
    }

    const bfs = (root, level) => {
        if (!root) {
            return;
        }

        if (!res[level - 1]) {
            res[level - 1] = [];
        }

        res[level - 1].push(root.val);

        bfs(root.left, level + 1);
        bfs(root.right, level + 1);
    };

    bfs(root, 1);

    return res;
};

const root = {
    val: 1,
    left: {
        val: 2,
        left: {
            val: 3,
            left: null,
            right: null
        },
        right: {
            val: 4,
            left: null,
            right: {
                val: 5,
                left: null,
                right: null
            }
        }
    },
    right: {
        val: 6,
        left: null,
        right: null
    }
};

/**
 * 测试用例：
 * console.log(levelOrder(root));
 */

/**
 * 本题核心：广度优先搜索
 * 
 * 因为需要记录层级，所以不能用队列的方式实现
 */